[Cocoa] Calculate quota based on disk space by default

szewai opened the pull request:
#### f7563eebd868af7f20acf3c2352c0243266e52d2
<pre>
[Cocoa] Calculate quota based on disk space by default
<a href="https://bugs.webkit.org/show_bug.cgi?id=255095">https://bugs.webkit.org/show_bug.cgi?id=255095</a>
rdar://107711361

Reviewed by Youenn Fablet.

Make quota calculated based on total disk space by setting default value of originQuotaRatio and totalQuotaRatio to be
non-null. Client could change it to null and fall back to old behavior (i.e. requesting quota to be value specified by
perOriginStorageQuota). The default values are picked based on current implementation of other browsers.
WebKitTestRunner still has old behavior as we have quite a few layout tests depending on the fixed quota value.

This patch also makes sure an exception is thrown if ratio argument is invalid (not in range [0, 1]).

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::donePrepareForEviction):
(WebKit::NetworkStorageManager::performEviction):
(WebKit::NetworkStorageManager::evictDataByTopOrigin):
(WebKit::NetworkStorageManager::getOriginDirectory):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.cpp:
(WKWebsiteDataStoreConfigurationHasOriginQuotaRatio):
(WKWebsiteDataStoreConfigurationClearOriginQuotaRatio):
(WKWebsiteDataStoreConfigurationHasTotalQuotaRatio):
(WKWebsiteDataStoreConfigurationClearTotalQuotaRatio):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration originQuotaRatio]):
(-[_WKWebsiteDataStoreConfiguration setOriginQuotaRatio:]):
(-[_WKWebsiteDataStoreConfiguration totalQuotaRatio]):
(-[_WKWebsiteDataStoreConfiguration setTotalQuotaRatio:]):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultOriginQuotaRatio):
(WebKit::WebsiteDataStore::defaultTotalQuotaRatio):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::defaultOriginQuotaRatio):
(WebKit::WebsiteDataStore::defaultTotalQuotaRatio):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBPersistence.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/StorageQuota.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::configureWebsiteDataStoreTemporaryDirectories):

Canonical link: <a href="https://commits.webkit.org/263250@main">https://commits.webkit.org/263250@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f7c0ce81851004bb0e3e132e31df133e83ae27b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3932 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4025 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4136 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5370 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4187 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3910 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4110 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4009 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3999 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3980 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4173 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3542 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5227 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1674 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3517 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5559 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3496 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3577 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4988 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3984 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3213 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3503 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3517 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/986 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3552 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3783 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -847,6 +847,16 @@ static NavigatingToAppBoundDomain schemeOrDomainIsAppBound(const String& host, c
+
+std::optional<double> WebsiteDataStore::defaultTotalQuotaRatio()
+{
+    return isFullWebBrowser() ? 0.80 : 0.20;
```
We tend to use meaningful names for constants like this.

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
```diff
@@ -715,7 +715,6 @@ HTTPServer server([connectionCount = 0] (Connection connection) mutable {
 {
     auto uuid = adoptNS([[NSUUID alloc] initWithUUIDString:@"68753a44-4d6f-1226-9c60-0050e4c00067"]);
     auto websiteDataStoreConfiguration = adoptNS([[_WKWebsiteDataStoreConfiguration alloc] initWithIdentifier:uuid.get()]);
-    EXPECT_NULL([websiteDataStoreConfiguration.get() originQuotaRatio]);
```
Can we add tests validating the default values?

youennf commented on Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.cpp:
```diff
@@ -203,3 +203,31 @@ void WKWebsiteDataStoreConfigurationSetPCMMachServiceName(WKWebsiteDataStoreConf
+{
+    std::optional<double> value;
+    if (ratio != -1.0)
+        value = ratio;
```
Would it make more sense to add a WKWebsiteDataStoreConfigurationClearTotalQuotaRatio instead of using a special -1 value?

szewai replied:
I use a special value -1 as I think it is easy to make `WKWebsiteDataStoreConfigurationGetOriginQuotaRatio` return -1  to indicate the value is not set.

Does it help if I change it to const?

youennf commented on Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
```diff
@@ -247,7 +251,22 @@ static bool isFullWebBrowser(const String& bundleIdentifier)
+        bool canOpenHTTP = [httpURLs containsObject:currentURL];
+        NSArray<NSURL *> *httpsURLs = [[NSWorkspace sharedWorkspace] URLsForApplicationsToOpenURL:[NSURL URLWithString:@"https:"]];
+        bool canOpenHTTPS = [httpsURLs containsObject:currentURL];
+        fullWebBrowser = canOpenHTTPS && canOpenHTTP;
```
This seems like a new definition, was it discussed somewhere?
Do we need this change in this patch, it seems TestWebKitAPI/WebKitTestRunner will anyway fall in the non browser cases?

szewai replied:
There is `return fullWebBrowser || isRunningTest(bundleIdentifier);` below, so we are treating test runners as browser.

youennf reviewed (commented):
LGTM, let's further discuss the full browser in Mac heuristic though.

youennf reviewed (approved):
LGTM
I would still prefer introducing WKWebsiteDataStoreConfigurationClearTotalQuotaRatio and WKWebsiteDataStoreConfigurationHasTotalQuotaRatio instead of -1.0

szewai committed:
[Cocoa] Calculate quota based on disk space by default
https://bugs.webkit.org/show_bug.cgi?id=255095
rdar://107711361

Reviewed by Youenn Fablet.

Make quota calculated based on total disk space by setting default value of originQuotaRatio and totalQuotaRatio to be
non-null. Client could change it to null and fall back to old behavior (i.e. requesting quota to be value specified by
perOriginStorageQuota). The default values are picked based on current implementation of other browsers.
WebKitTestRunner still has old behavior as we have quite a few layout tests depending on the fixed quota value.

This patch also makes sure an exception is thrown if ratio argument is invalid (not in range [0, 1]).

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::donePrepareForEviction):
(WebKit::NetworkStorageManager::performEviction):
(WebKit::NetworkStorageManager::evictDataByTopOrigin):
(WebKit::NetworkStorageManager::getOriginDirectory):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.cpp:
(WKWebsiteDataStoreConfigurationHasOriginQuotaRatio):
(WKWebsiteDataStoreConfigurationClearOriginQuotaRatio):
(WKWebsiteDataStoreConfigurationHasTotalQuotaRatio):
(WKWebsiteDataStoreConfigurationClearTotalQuotaRatio):
* Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration originQuotaRatio]):
(-[_WKWebsiteDataStoreConfiguration setOriginQuotaRatio:]):
(-[_WKWebsiteDataStoreConfiguration totalQuotaRatio]):
(-[_WKWebsiteDataStoreConfiguration setTotalQuotaRatio:]):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultOriginQuotaRatio):
(WebKit::WebsiteDataStore::defaultTotalQuotaRatio):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::defaultOriginQuotaRatio):
(WebKit::WebsiteDataStore::defaultTotalQuotaRatio):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IndexedDBPersistence.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/StorageQuota.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::configureWebsiteDataStoreTemporaryDirectories):

Canonical link: https://commits.webkit.org/263250@main
